### PR TITLE
Refactor and restructure anomalib.data

### PR DESCRIPTION
**Pull request by samet-akcay**
## 📝 Description

- Split Torch datasets and Lightning Datamodules into `datasets` and `datamodules`
- Move `dataclasses` into `data` 

## ✨ Changes

Select what type of change your PR is:

- [ ] 🐞 Bug fix (non-breaking change which fixes an issue)
- [x] 🔨 Refactor (non-breaking change which refactors the code base)
- [ ] 🚀 New feature (non-breaking change which adds functionality)
- [ ] 💥 Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] 📚 Documentation update
- [ ] 🔒 Security update

## ✅ Checklist

Before you submit your pull request, please make sure you have completed the following steps:

- [ ] 📋 I have summarized my changes in the [CHANGELOG](https://github.com/openvinotoolkit/anomalib/blob/main/CHANGELOG.md) and followed the guidelines for my type of change (skip for minor changes, documentation updates, and test enhancements).
- [ ] 📚 I have made the necessary updates to the documentation (if applicable).
- [ ] 🧪 I have written tests that support my changes and prove that my fix is effective or my feature works (if applicable).

For more information about code review checklists, see the [Code Review Checklist](https://github.com/openvinotoolkit/anomalib/blob/main/docs/source/markdown/guides/developer/code_review_checklist.md).

### CODE REVIEW

**Review by djdameln — CHANGES_REQUESTED** [review]
As discussed, I'm in favor of this new structure. It seems the `datasets` submodule is missing though.

**Comment by samet-akcay** [comment]
> As discussed, I'm in favor of this new structure. It seems the `datasets` submodule is missing though.

no idea what's wrong though. Tests locally pass 🙃

File: src/anomalib/data/datamodules/base/__init__.py [hunk 1]
```diff
@@ -0,0 +1,9 @@
+"""Base Anomalib Data Modules."""
+
+# Copyright (C) 2022 Intel Corporation
```
Comment (ashwinvaidya17) [inline src/anomalib/data/datamodules/base/__init__.py:3]:
```suggestion
# Copyright (C) 2022-2024 Intel Corporation
```

**Review by ashwinvaidya17 — APPROVED** [review]
So from what I understand, the intent of the PR is to split data into dataclasses and datamodules?

**Comment by samet-akcay** [comment]
> So from what I understand, the intent of the PR is to split data into dataclasses and datamodules?

and also datasets. Overall, data will contain the dataclasses (item, batch etc.), datasets (torch) and datamodules (lightning)

**Review by djdameln — APPROVED** [review]

**Pull request merged**